Physics interpolation docs (#5801)
godotengine/godot-docs: tutorials/physics/interpolation/advanced_physics_interpolation.rst

.. _doc_advanced_physics_interpolation:

> Advanced Physics Interpolation

Reviewer: ```suggestion
Advanced physics interpolation
```

==============================

Although the previous instructions will give satisfactory results in a lot of games, in some cases you will want to go a stage further to get the best possible results and the smoothest possible experience.

Exceptions to automatic physics interpolation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Even with physics interpolation active, there may be some local situations where you would benefit from disabling automatic interpolation for a :ref:`Node<class_Node>` (or branch of the :ref:`SceneTree<class_SceneTree>`), and have the finer control of performing interpolation manually.

This is possible using the :ref:`Node.set_physics_interpolated<class_Node_method_set_physics_interpolated>` function which is present in all Nodes. If you for example, set this interpolated flag to false for a Node, the children will recursively also be affected. This means you can easily disable interpolation for an entire subscene.

The most common situation where you may want to perform your own interpolation is Cameras.

Cameras
^^^^^^^

> In many cases a :ref:`Camera<class_Camera>` can just use automatic interpolation just like any other node. However, for best results, especially at low physics tick rates, it is recommended that you take a manual approach to Camera interpolation.

Reviewer: ```suggestion
In many cases, a :ref:`Camera<class_Camera>` can use automatic interpolation just like any other node. However, for best results, especially at low physics tick rates, it is recommended that you take a manual approach to Camera interpolation.
```

> This is because viewers are very sensitive to Camera movement. A Camera that realigns slightly every, say 1/10th of a second at 10tps tick rate will often be noticeable, and you can get a much smoother result by moving the Camera each frame in ``_process``, and following an interpolated target manually.

Reviewer: ```suggestion
This is because viewers are very sensitive to Camera movement. For instance, a Camera that realigns slightly every 1/10th of a second (at 10tps tick rate) will often be noticeable. You can get a much smoother result by moving the Camera each frame in ``_process``, and following an interpolated target manually.
```

Manual Camera interpolation
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> **Ensure Camera is using global space**

Reviewer: ```suggestion
**Ensure the Camera is using global coordinate space instead of the default local space.**
```

Author: Omitted the "instead of default local space", because local space can still be the same as global space (if the parent is identity or it is set as top level).

A typical example of a custom approach is to use the ``look_at`` function in the Camera every frame in ``_process()`` to look at a target node (for example the player).

But there is a problem. If we use the traditional ``get_global_transform()`` on a Camera "target" Node, this transform will only focus the Camera on the target *at the current physics tick*. This is *not* what we want, as the Camera will jump about on each physics tick as the target moves. Even though the Camera may be updated each frame, this does not help give smooth motion if the *target* is only changing each physics tick.

get_global_transform_interpolated()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

What we really want to focus the Camera on, is not the position of the target on the physics tick, but the *interpolated* position, i.e. the position at which the target will be rendered.

We can do this using the :ref:`Spatial.get_global_transform_interpolated<class_Spatial_method_get_global_transform_interpolated>` function. This acts exactly like getting :ref:`Spatial.global_transform<class_Spatial_property_global_transform>` but it gives you the *interpolated* transform (during a ``_process()`` call).

> .. important:: ``get_global_transform_interpolated()`` should only be used once or twice for special cases such as Cameras. It should **not** be used all over the place in your code (both for performance reasons, and to give correct gameplay).
> 
> .. note:: Aside from exceptions like the Camera, in most cases your game logic should be in ``_physics_process()`` and you should be calling ``get_global_transform()`` or ``get_transform()``, which will give the current physics transform (in global or local space respectively), which is usually what you will want for gameplay code.

Reviewer: ```suggestion
.. important:: ``get_global_transform_interpolated()`` should only be used once or twice for special cases such as Cameras. It should **not** be used all over the place in your code (both for performance reasons, and to give correct gameplay).

.. note:: Aside from exceptions like the Camera, in most cases, your game logic should be in ``_physics_process()``. In game logic, you should be calling ``get_global_transform()`` or ``get_transform()``, which will give the current physics transform (in global or local space respectively), which is usually what you will want for gameplay code.
```

Example manual Camera script
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Here is an example of a simple fixed Camera which follows an interpolated target:

.. code-block:: python

	extends Camera
		
	# Node that the camera will follow
	var _target
		
	# We will smoothly lerp to follow the target
	# rather than follow exactly
	var _target_pos : Vector3 = Vector3()
		
	func _ready() -> void:
		# Find the target node
		_target = get_node("../Player")
		
		# Turn off automatic physics interpolation for the Camera,
		# we will be doing this manually
		set_physics_interpolated(false)
		
	func _process(delta: float) -> void:
		# Find the current interpolated transform of the target
		var tr : Transform = _target.get_global_transform_interpolated()
		
		# Provide some delayed smoothed lerping towards the target position 
		_target_pos = lerp(_target_pos, tr.origin, min(delta, 1.0))
		
		# Fixed camera position, but it will follow the target
		look_at(_target_pos, Vector3(0, 1, 0))

Mouse look
^^^^^^^^^^

> Mouse look is a very common way of controlling Cameras. But there is a problem. Unlike keyboard input, the mouse is not polled at each physics tick. New mouse move events come in continuously, and the Camera will be expected to react and follow these on the next frame, rather than waiting until the next physics tick.

Author: Ah I'll try and reword this. I'm meaning you might typically do an `Input.is_action_pressed()` in `_physics_process()` when using interpolation, but the mouse input is more in need of responding to immediately.

> Mouse look is a very common way of controlling Cameras. But there is a problem. Unlike keyboard input, the mouse is not polled at each physics tick. New mouse move events come in continuously, and the Camera will be expected to react and follow these on the next frame, rather than waiting until the next physics tick.

Author: Have reworded this, you may want to check. :+1:

> In this situation it can be better to disable physics interpolation for the Camera node (using :ref:`Node.set_physics_interpolated<class_Node_method_set_physics_interpolated>`) and directly apply the mouse input to the Camera rotation, rather than apply it in ``_physics_process``.

Reviewer: ```suggestion
In this situation, it can be better to disable physics interpolation for the Camera node (using :ref:`Node.set_physics_interpolated<class_Node_method_set_physics_interpolated>`) and directly apply the mouse input to the Camera rotation, rather than apply it in ``_physics_process``.
```

Sometimes, especially with Cameras, you will want to use a combination of interpolation and non-interpolation:

* A first person camera may position the camera at a player location (perhaps using :ref:`Spatial.get_global_transform_interpolated<class_Spatial_method_get_global_transform_interpolated>`), but control the Camera rotation from mouse look *without* interpolation.
* A third person camera may similarly determine the look at (target location) of the camera using :ref:`Spatial.get_global_transform_interpolated<class_Spatial_method_get_global_transform_interpolated>`, but position the camera using mouse look *without* interpolation.

There are many permutations and variations of Camera types, but it should be clear that in many cases, disabling automatic physics interpolation and handling this yourself can give a better result.

Disabling interpolation on other nodes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Although Cameras are the most common example, there are a number of cases when you may wish other nodes to control their own interpolation, or be non-interpolated. Consider for example, a player in a top view game whose rotation is controlled by mouse look. Disabling physics rotation allows the player rotation to match the mouse in realtime.
> 

Reviewer: ```suggestion
Although Cameras are the most common example, there are a number of cases when you may wish other nodes to control their own interpolation, or be non-interpolated. Consider for example, a player in a top view game whose rotation is controlled by mouse look. Disabling physics rotation allows the player rotation to match the mouse in real-time.
```
